server/topic.go

```diff
@@ -1,43 +1,75 @@
 package server
 
 import (
-	"heckel.io/ntfy/log"
 	"math/rand"
 	"sync"
+	"time"
+
+	"heckel.io/ntfy/log"
 )
 
 // topic represents a channel to which subscribers can subscribe, and publishers
 // can publish a message
 type topic struct {
 	ID          string
-	subscribers map[int]subscriber
+	subscribers map[int]topicSubscription
+	lastUnsub   topicSubscription
 	mu          sync.Mutex
 }
 
 // subscriber is a function that is called for every new message on a topic
+type topicSubscription struct {
+	f         subscriber
+	v         *visitor
+	unsubTime time.Time
+}
 type subscriber func(v *visitor, msg *message) error
 
 // newTopic creates a new topic
 func newTopic(id string) *topic {
 	return &topic{
 		ID:          id,
-		subscribers: make(map[int]subscriber),
+		subscribers: make(map[int]topicSubscription),
 	}
 }
 
+// need a better name for bill?
+// Returns nil, nil for non-UP topics
+// returns visitor, nil for active UP topics
+// returns nil, err for inactive UP topics
+func (t *topic) getBillee() *visitor {
+	//get a pseudo random visitor???
+	for _, this_subscriber := range t.subscribers {
+		return this_subscriber.v
+	}
+	// what if someone unsubscribed and DOESNT want their sub to count against them anymore, maybe the app server lost sync and will keep on sending stuff
+	// I guess they suffer for unifiedPushSubscriptionDuration?
+
+	// if lastunsub v exists, and the time since it was unsubbed is longer than our limit, it should not exist
+	if t.lastUnsub.v != nil && time.Since(t.lastUnsub.unsubTime) > unifiedPushSubscriptionDuration {
+		t.lastUnsub.v = nil
+	}
```

Comment:
That. How do we know whether the last subscriber wants to cache messages (gone temporarily) or not (unsubscribed permanently)?

Reply:
I don't understand what's happening here. Sorry.

Reply:
Messages are being billed to the last visitor (either someone currently subscribed or the last person to unsubscribe). This bill-the-last-unsusbscriber behavior remains for `unifiedPushSubscriptionDuration`.

Let's say `user:KM` is subscribed to ntfy.sh/upABCD, and `unifiedPushSubscriptionDuration := 10 * time.Hour`

**Scenario 1:**
06:00 `user:KM`'s internet went out and they unsubscribed from ntfy.sh/upABCD
06:15: ntfy.sh receives a message at upABCD. It is stored and billed to `KM`
06:30 `user:KM` is online again. The missed message is delivered and everyone's happy.

**Scenario 2:**
06:00 `user:KM`'s internet went out and they unsubscribed from ntfy.sh/upABCD
06:15: ntfy.sh receives a message at upABCD. It is stored and billed to `KM`
15:59: ntfy.sh receives a message at upABCD. It is stored and billed to `KM`
16:01: ntfy.sh receives a message at upABCD. `t.lastUnsub.v` is set to `nil`. The message is rejected (412 or whatever else).
23:59 `user:KM` is online again. They missed some messages, but it's not *that* big of a deal.

**Scenario 3:**
06:00 `user:KM`'s intentionally unsubscribed from ntfy.sh/upABCD. They don't want upABCD messages anymore, but the app server keeps sending them.
06:15: ntfy.sh receives a message at upABCD. It is stored and billed to `KM`
15:59: ntfy.sh receives a message at upABCD. It is stored and billed to `KM`
16:01: ntfy.sh receives a message at upABCD. `t.lastUnsub.v` is set to `nil`. The message is rejected (412 or whatever else).
`user:KM` is angry that 2 (or 200 or 2000) messages from their rate limit were eaten up by this random app server.